datalad/interface/run.py

```diff
@@ -458,10 +490,79 @@ def format_command(command, **kwds):
     return sfmt.format(command, **kwds)
 
 
-# This helper function is used to add the rerun_info argument.
+def _execute_command(command, pwd, expected_exit=None):
+    from datalad.cmd import Runner
+
+    exc = None
+    cmd_exitcode = None
+    runner = Runner(cwd=pwd)
+    try:
+        lgr.info("== Command start (output follows) =====")
+        runner.run(
+            command,
+            # immediate output
+            log_online=True,
+            # not yet sure what we should do with the command output
+            # IMHO `run` itself should be very silent and let the command talk
+            log_stdout=False,
+            log_stderr=False,
+            expect_stderr=True,
+            expect_fail=True,
+            # TODO stdin
+        )
+    except CommandError as e:
+        # strip our own info from the exception. The original command output
+        # went to stdout/err -- we just have to exitcode in the same way
+        exc = e
+        cmd_exitcode = e.code
+
+        if expected_exit is not None and expected_exit != cmd_exitcode:
+            # we failed in a different way during a rerun.  This can easily
+            # happen if we try to alter a locked file
+            #
+            # TODO add the ability to `git reset --hard` the dataset tree on failure
+            # we know that we started clean, so we could easily go back, needs gh-1424
+            # to be able to do it recursively
+            raise exc
+
+    lgr.info("== Command exit (modification check follows) =====")
+    return cmd_exitcode or 0, exc
+
+
 def run_command(cmd, dataset=None, inputs=None, outputs=None, expand=None,
                 explicit=False, message=None, sidecar=None,
-                rerun_info=None, rerun_outputs=None):
+                extra_info=None,
+                rerun_info=None, rerun_outputs=None,
+                inject=False):
+    """Run `cmd` in `dataset` and record the results.
+
+    `Run.__call__` is a simple wrapper over this function. Aside from backward
+    compatibility kludges, the only difference is that `Run.__call__` doesn't
+    expose all the parameters of this function. The unexposed parameters are
+    listed below.
+
+    Parameters
+    ----------
+    extra_info : dict, optional
+        Additional information to dump with the json run record. Any value
+        given here will take precedence over the standard run key. Warning:
+        Callers should try to use fairly specific key names to avoid collisions
+        with future keys added by `run`.
```

Comment:
minor: I wonder if we should even encourage adding additional information in subsections (e.g. `niceman` to contain any relevant/added field, ...) instead of flooding this record at the top level?

Reply:
Sounds like a good idea, yes.